### ISSUE TITLE
[YSQL] A single-key write followed by full-table scan invariant check does not detect a conflict in serializable mode

### ISSUE THREAD

**Issue by mbautin** [yugabyte/yugabyte-db#1572]
This is a special case of what @bllewell discovered in #1498 ("one or two admins" test).

We would expect a conflict to be detected in this Java test:

```java
  @Test
  public void testOneOrTwoAdmins() throws Exception {
    Statement setupStatement = connection.createStatement();
    setupStatement.execute(
        "create table staff(" +
        "  name text constraint staff_name primary key, " +
        "  job  text constraint staff_job_id_nn not null, " +
        "            constraint staff_job_id_chk check (job in ( " +
        "              'Manager', "+
        "              'Admin', " +
        "              'Sales', " +
        "              'Marketing', " +
        "              'Developer'))" +
        "  );");
    setupStatement.execute(
        "insert into staff(name, job) values" +
        "  ('Mary',  'Manager')," +
        "  ('Susan', 'Developer')," +
        "  ('Helen', 'Sales')," +
        "  ('Bill',  'Marketing')," +
        "  ('Fred',  'Sales')," +
        "  ('John',  'Admin');");
    final IsolationLevel isolation = IsolationLevel.SERIALIZABLE;
    Connection connection1 = createConnection(isolation, AutoCommit.DISABLED);
    Statement statement1 = connection1.createStatement();
    
    Connection connection2 = createConnection(isolation, AutoCommit.DISABLED);
    Statement statement2 = connection2.createStatement();

    statement1.execute("insert into staff(name, job) values('Alice', 'Admin');");
    statement2.execute("insert into staff(name, job) values('Bert', 'Admin');");

    LOG.info("statement1 assertion: " + executeOneOrTwoAdminsAssertion(statement1));
    LOG.info("statement2 assertion: " + executeOneOrTwoAdminsAssertion(statement2));

    connection1.commit();
    connection2.commit();
  }
```

And the invariant checking function is like this:
```
  private String executeOneOrTwoAdminsAssertion(Statement statement) throws SQLException {
      ResultSet result = statement.executeQuery("select * from staff where job = 'Admin'");
      int count = 0;
      while (result.next()) {
        count++;
      }
      if (1 <= count && count <= 2)
        return "passed";
      else
        return "failed";
  }
```
or could be just `SELECT count(*) FROM staff WHERE job = 'Admin'`.

**Comment by mbautin** [comment]
However, if the assertion check is moved **before** the write statements, the conflict is successfully detected:
```
    LOG.info("statement1 assertion: " + executeOneOrTwoAdminsAssertion(statement1));
    LOG.info("statement2 assertion: " + executeOneOrTwoAdminsAssertion(statement2));

    statement1.execute("insert into staff(name, job) values('Alice', 'Admin');");
    statement2.execute("insert into staff(name, job) values('Bert', 'Admin');");
```

**Comment by bllewell** [comment]
But you have to wait until you’ve done the intended changes before checking the data rule assertion.

On 19-Jun-2019, at 17:16, Mikhail Bautin <notifications@github.com> wrote:

However, if the assertion check is moved before the write statements, the conflict is successfully detected:

    LOG.info("statement1 assertion: " + executeOneOrTwoAdminsAssertion(statement1));
    LOG.info("statement2 assertion: " + executeOneOrTwoAdminsAssertion(statement2));

    statement1.execute("insert into staff(name, job) values('Alice', 'Admin');");
    statement2.execute("insert into staff(name, job) values('Bert', 'Admin');");

**Comment by mbautin** [comment]
The issue seems to be (and @spolitov confirmed this is the bug) that when we send a read request to a tablet server, the YBClient layer only sends transaction metadata (status tablet, isolation level, etc.) to that tablet server if it thinks that the tablet has not yet persisted the metadata in its provisional records DB. However, the case of only transaction id having been sent to the tablet in the read request (for the second request to the same tablet, in this case the full-table scan read request) was being handled incorrectly, so the read request could not detect the isolation level of the transaction and went through the repeatable read path to handle the transaction. For tablets where no rows were previously written by the same transaction the transaction metadata was sent by the client correctly.

**Comment by bmatican** [comment]
Fixed in #1498
